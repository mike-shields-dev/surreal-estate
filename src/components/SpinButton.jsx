import React from "react";
import PropTypes from "prop-types";

const SpinButton = ({ field, min, onChange, required, step, units, value }) => {
  return (
    <label htmlFor={field}>
      {field} {units}
      <input
        required={required}
        id={field}
        min={min}
        name={field}
        onChange={onChange}
        step={step}
        type="number"
        value={value}
      />
    </label>
  );
};

SpinButton.defaultProps = {
  min: null,
  units: null,
  step: null,
  required: false,
};

SpinButton.propTypes = {
  field: PropTypes.string.isRequired,
  min: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  required: PropTypes.bool,
  step: PropTypes.string,
  units: PropTypes.string,
  value: PropTypes.number.isRequired,
};

export default SpinButton;
