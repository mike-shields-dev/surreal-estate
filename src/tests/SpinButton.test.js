import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import SpinButton from "../components/SpinButton";

const props = {
  field: "bedrooms",
  min: "0",
  onChange: jest.fn(),
  step: "1",
  units: "$",
  value: "1",
};

const renderSpinButton = () => render(<SpinButton {...props} />);

describe("SpinButton", () => {
  it("matches snapshot", () => {
    const { asFragment } = renderSpinButton();

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a number input with the given props and an associated label", () => {
    renderSpinButton();
    const numberInput = screen.getByRole("spinbutton", {
      name: `${props.field} ${props.units}`,
    });

    expect(numberInput).toHaveAttribute("id", props.field);
    expect(numberInput).toHaveAttribute("name", props.field);
    expect(numberInput).toHaveAttribute("min", props.min);
    expect(numberInput).toHaveAttribute("step", props.step);
    expect(numberInput).toHaveAttribute("type", "number");
    expect(numberInput).toHaveAttribute("value", props.value);
  });

  it("renders a number input", () => {
    renderSpinButton();

    expect(screen.getByRole("spinbutton")).toBeInTheDocument();
  });

  it("invokes the onChange callback", async () => {
    renderSpinButton();

    fireEvent.change(screen.getByRole("spinbutton"), {
      target: { value: props.value + 1 },
    });

    expect(props.onChange).toHaveBeenCalled();
  });
});
