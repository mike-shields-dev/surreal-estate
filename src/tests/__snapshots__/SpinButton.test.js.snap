// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinButton matches snapshot 1`] = `
<DocumentFragment>
  <label
    for="bedrooms"
  >
    bedrooms $
    <input
      id="bedrooms"
      min="0"
      name="bedrooms"
      step="1"
      type="number"
      value="1"
    />
  </label>
</DocumentFragment>
`;
